dilated-file-transfer.md

```diff
@@ -0,0 +1,484 @@
+The offering side MAY show the "reason" string to the user.
+
+When the offering side gets an `OfferAccept` message it begins streaming the file over the already-opened subchannel.
+When completed, the subchannel is closed.
+
+That is, the offering side always initiates the open and close of the corresponding subchannel.
+
+For the example above, the sending side determines whether it should stream data (e.g. in mode "yes" it should start right away, otherwise wait for an `OfferAccept`).
+
+It will then send messages in this order:
+* a kind `1` `FileOffer(filename="README")`
+* a kind `5` data with 65 bytes of payload
+* a kind `1` `FileOffer(filename="hello.py")`
+* a kind `5` data with 100 bytes of payload
+* close the subchannel
+
+Messages of kind `5` ("file data bytes") consist solely of file data.
+A single data message MUST NOT exceed 65536 bytes (65KiB) including the single byte for "kind" (so 65535 maximum payload bytes).
+Applications are free to choose how to fragment the file data so long as no single message is bigger than 65536 bytes.
+A good default to choose in 2022 is 16KiB (2^14 - 1 payload bytes)
+
+    XXX: what is a good default? Dilation doesn't give guidance either...
+
+When sending a `DirectoryOffer` each individual file is preceded by a `FileOffer` message.
+However the rules about "maybe wait for reply" no longer exist; that is, all file data MUST be immediately sent (the `FileOffer`s serve as a header).
+
+See examples down below, after "Discussion".
+
+
+## Discussion and Open Questions {#discussion}
+
+* Overall versioning considerations
+
+Versioning is hard.
+
+The existing file-transfer protocol does not include versioning.
+Luckily, the mailbox protocol _itself_ allows for "application versioning" messages; this gives us an "out" here, effectively letting us send a "T minus 1" message via the "application version" dict.
+
+We do not have another such escape hatch (i.e. "T minus 2"), so if we get it wrong (again) then we potentially have a painful upgrade path.
+
+Currently, a file-transfer peer that sends zero version data is assumed to be "classic".
+A file-transfer peer supporting Dilation and this new protocol sends `"transfer": {...}` as per the  "Version Negotiation" section above.
+We include a `"version"` key in that dict so that this transfer protocol may be extended in the future (see "Protocol Expansion Exercises" below).
+
+Additionally, a `"features"` key is included in that information.
+Although related, this is somewhat orthogonal to "versions".
+That is, a peer may _know how to parse_ some (newer) version of this protocol but may still wish to _not_ support (or use) a particular feature.
+
+
+* message encoding
+
+While `msgpack` is mentioned above, there are several other binary-supporting libraries worth considering.
+These are (in no particular order) at least: CBOR or flatbuffers or protocolbuffers or cap'n'proto
```

Comment:
One requirement for the serialization format is that it is self-describing, i.e. no external schema is required in order to serialize and deserialize it. As far as I know, this rules out protobuf and friends. So I think viable candidates are msgpack, CBOR and BSON.

Reply:
Why is that a requirement?

Reply:
Because otherwise both sides need to agree on an *exact* scheme to encode their messages with. With self-describing formats, one can freely add new fields on one side and having the other ignore them if unsupported. Problematic mismatches can be caught by the parser more easily, instead of causing (sometimes silent) data corruption.

Reply:
We are specifying the exact scheme, though.

Reply:
(or, what messages do you think we might expand without an `"abilities"` or protocol-version change? and why?)

Reply:
Hm, now that we have some stronger versioning than in some of the older Magic Wormhole protocol parts, maybe this would be feasible. Because previously, there always was this "ignore unknown fields" for backwards compatibility, but we don't need that anymore (except for the versions message, but that's JSON anyways).

I haven't looked at those formats because of the reasons above previously, but maybe it's worth spending a few hours making sure none of them suits our use case better than msgpack.

Reply:
We don't need it for the version message either (now that the proposal shapes it like `transfer-v1` etc).

Reply:
How is this supposed to work? How is a client going to know the schema for fields that are going to be added in the future?

Reply:
If you rely on the version in the tag (as you suggested) then you have to make a whole new dict for each version supported, like `"transfer-v2": {"new-tag": 1}` so whenever you change the version information it's a new protocol version.

With the version "inside" then you do indeed have to say "ignore new keys"

Reply:
Anyway it's JSON anyway like you said, so ..